### PR TITLE
Run tests with multiple configurable FRR containers

### PR DESCRIPTION
**Pull request by oribon**
Currently the tests run using one external BGP peer.
Running multiple external BGP peers allows us to cover
more scenarios in parallel such as internal/external BGP
and multiple MyASNs. In addition, doing it in a configurable
way will allow us to extend the tests to cover more scenarios
in the future by shuffling configuration for the speaker
and the external containers that serve as BGP peers.

Signed-off-by: Ori Braunshtein <obraunsh@redhat.com>

### CODE REVIEW

File: e2etest/bgp_test.go [hunk 1]
```diff
@@ -21,72 +21,107 @@ package e2e
...
-			}
-		} else {
-			frrTestConfigDir, err = ioutil.TempDir("", "frr-conf")
-			framework.ExpectNoError(err)
-			err = frr.StartContainer(frrContainer, frrTestConfigDir)
-			framework.ExpectNoError(err)
+		frrContainers = make([]*frrcontainer.Frr, 0)
+		g := new(errgroup.Group)
+		for _, conf := range containersConf {
+			conf := conf
```
Comment (sabinaaledort) [inline e2etest/bgp_test.go:94]: seems odd

File: e2etest/pkg/frr/container/container.go [hunk 2]
```diff
@@ -19,20 +20,78 @@ const (
...
+	configDir      string
+	NeighborConfig config.NeighborConfig
+	RouterConfig   config.RouterConfig
+	Ipv4           string
+	Ipv6           string
+}
+
+// New creates a new FRR container on the host and returns the corresponding *Frr.
+// A situation where a non-nil container and an error are returned is possible.
+func New(name string, nc config.NeighborConfig, rc config.RouterConfig, network string, hostipv4 string, hostipv6 string) (*Frr, error) {
```
Comment (sabinaaledort) [inline e2etest/pkg/frr/container/container.go:41]: maybe we can find a better name than new? when reading the test code it wasn't clear for me that this function is staring the container

File: e2etest/pkg/frr/container/container.go [hunk 3]
```diff
@@ -42,50 +101,38 @@ func StartContainer(containerName string, testDirName string) error {
 }
 
 // Returns the IPv4 and IPv6 addresses of the container.
```
Comment (sabinaaledort) [inline e2etest/pkg/frr/container/container.go:103]: maybe we should update the comment as this function no longer returns the ips but setting them

File: e2etest/pkg/frr/container/container.go [hunk 4]
```diff
@@ -111,9 +158,24 @@ func UpdateContainerVolumePermissions(containerName string) error {
...
+	out, err := exec.Command("docker", "exec", c.Name, "sh", "-c", cmd).CombinedOutput()
+	if err != nil {
+		return errors.Wrapf(err, "Failed to change %s container volume permissions. %s", c.Name, string(out))
+	}
+
+	return nil
+}
+
+// Updating the BGP config file.
+func (c *Frr) UpdateBGPConfigFile(bgpConfig string) error {
```
Comment (sabinaaledort) [inline e2etest/pkg/frr/container/container.go:170]: why did you move the function? might look confusing in the log

File: e2etest/bgp_test.go (hunk 1 above)
Comment (oribon) [reply e2etest/bgp_test.go:94]: https://golang.org/doc/faq#closures_and_goroutines

File: e2etest/pkg/frr/container/container.go (hunk 2 above)
Comment (oribon) [reply e2etest/pkg/frr/container/container.go:41]: What do you suggest? For me it feels new container = start a container with additional stuff

File: e2etest/pkg/frr/container/container.go (hunk 3 above)
Comment (oribon) [reply e2etest/pkg/frr/container/container.go]: done

File: e2etest/pkg/frr/container/container.go (hunk 4 above)
Comment (oribon) [reply e2etest/pkg/frr/container/container.go]: done

File: e2etest/bgp_test.go [hunk 5]
```diff
@@ -21,72 +21,107 @@ package e2e
...
+			},
+		},
+		{
+			name: frrEBGP,
+			nc: frrconfig.NeighborConfig{
+				ASN: EBGPAsn,
+			},
+			rc: frrconfig.RouterConfig{
+				ASN:     IBGPAsn,
+				BGPPort: 180,
```
Comment (msherif1234) [inline e2etest/bgp_test.go:85]: hmm this BGPPort is different than the one /neighbor don't we need to keep this at 0 as we have today and just change the /neighbor port if configured ?
`When port number is 0, that means do not listen bgp port.`

File: e2etest/bgp_test.go [hunk 6]
```diff
@@ -95,6 +130,15 @@ var _ = ginkgo.Describe("BGP", func() {
 	ginkgo.Context("type=Loadbalancer", func() {
 		var allNodes *corev1.NodeList
 		ginkgo.BeforeEach(func() {
+			var peers []peer
+			for _, c := range frrContainers {
+				peers = append(peers, peer{
+					Addr:  c.Ipv4,
+					ASN:   c.RouterConfig.ASN,
+					MyASN: c.NeighborConfig.ASN,
```
Comment (msherif1234) [inline e2etest/bgp_test.go:138]: did u get the above swapped ?

File: e2etest/config/frr/daemons [hunk 7]
```diff
@@ -38,7 +38,7 @@ vrrpd=no
 #
 vtysh_enable=yes
 zebra_options="  -A 127.0.0.1 -s 90000000"
-bgpd_options="   -A 127.0.0.1"
+bgpd_options="   -A 127.0.0.1 -p {{.BGPPort}}"
```
Comment (msherif1234) [inline e2etest/config/frr/daemons:41]: I don't think we need to set the above ?

File: e2etest/pkg/frr/config/config.go [hunk 8]
```diff
@@ -18,30 +18,30 @@ import (
 // BGP router config.
 const bgpConfigTemplate = `
 router bgp {{.ASN}}
+  no bgp ebgp-requires-policy
```
Comment (msherif1234) [inline e2etest/pkg/frr/config/config.go:21]: curious what issues u faced to do the above I did bgp peering for both iBGP and eBGP w/o the need to this
I thought because we have `no bgp network import-check`, also I assume we are running as datacenter hence its off by default ? `This is enabled by default for the traditional configuration and turned off by default for datacenter configuration.`

File: e2etest/pkg/frr/config/config.go [hunk 9]
```diff
@@ -105,3 +103,34 @@ func SetBGPConfig(testDirName string, config string) error {
...
+		BGPPort: rc.BGPPort,
+	}
+
+	err = tpl.Execute(f, info)
+	if err != nil {
+		return errors.Wrapf(err, "Failed to update %s", path)
+	}
+
+	return nil
+}
```
Comment (msherif1234) [inline e2etest/pkg/frr/config/config.go:136]: why we have this function here and its used only by test ?

File: e2etest/pkg/frr/container/container.go [hunk 10]
```diff
@@ -19,20 +20,78 @@ const (
...
 	}
 
+	err = config.SetDaemonsConfig(testDirName, rc)
+	if err != nil {
+		return err
+	}
+
 	volume := fmt.Sprintf("%s:%s", testDirName, frrMountPath)
-	out, err := exec.Command("docker", "run", "-d", "--privileged", "--network", "kind", "--rm", "--name", containerName,
+	out, err := exec.Command("docker", "run", "-d", "--privileged", "--network", network, "--rm", "--name", containerName,
```
Comment (msherif1234) [inline e2etest/pkg/frr/container/container.go:94]: why do we need to run as privileged ?

**Comment by msherif1234** [comment]
How did u test multiple MyASN w/oi using view feature in #1015 ?

File: e2etest/bgp_test.go (hunk 6 above)
Comment (oribon) [reply e2etest/bgp_test.go:138]: From the speaker's point of view he's the container's neighbor (MyASN should come from neighborconfig) and ASN,Addr corresponds to the RouterConfig and container's IP.

File: e2etest/bgp_test.go (hunk 5 above)
Comment (oribon) [reply e2etest/bgp_test.go:85]: This also relates to https://github.com/metallb/metallb/pull/1018#discussion_r740222489
that's the BGPPort that the external container will listen on so it is set here and reflected in the peer (c.RouterConfig.BGPPort) so the speaker would target the correct port.

File: e2etest/config/frr/daemons (hunk 7 above)
Comment (oribon) [reply e2etest/config/frr/daemons:41]: ^^ this is to configure FRR to listen on a specified port, useful for checking the configmap's port option AND configuring 2 containers on the hostnetwork.

File: e2etest/pkg/frr/config/config.go (hunk 9 above)
Comment (oribon) [reply e2etest/pkg/frr/config/config.go:136]: what do you mean?

File: e2etest/pkg/frr/container/container.go (hunk 10 above)
Comment (oribon) [reply e2etest/pkg/frr/container/container.go:94]: that wasn't part of my commit :sweat_smile:

File: e2etest/bgp_test.go (hunk 5 above)
Comment (msherif1234) [reply e2etest/bgp_test.go:85]: what if we change the port later `neighbor 1.1.1.1 port 200` and what will happen if we have 
`neighbor 1.1.1.1 port 100`
`nieghbor 1.1.1.2 port 110`
which one will be used for BGPPort ?

File: e2etest/pkg/frr/config/config.go (hunk 9 above)
Comment (msherif1234) [reply e2etest/pkg/frr/config/config.go:136]: just curious why you add it here not in the testfile itself

**Comment by oribon** [comment]
> How did u test multiple MyASN w/oi using view feature in #1015 ?

It runs on the native bgp stack - it shows that it works with multiple MyASNs hence we'd want to have these tests pass when FRR with views land. The external containers here don't need to use views (they're using a single ASN).

File: e2etest/bgp_test.go (hunk 6 above)
Comment (msherif1234) [reply e2etest/bgp_test.go:138]: I understand ur explaintion IMO this will be source of continous confusion we should just consider them as two separate routers and use same syntax its known remote-as of neghibor has the match router instance

File: e2etest/pkg/frr/container/container.go (hunk 10 above)
Comment (msherif1234) [reply e2etest/pkg/frr/container/container.go:94]: probably because the routes are added to kernel however with view usage we don't install routes in the kernel maybe we can remove that later i will check it out

Comment (oribon) [reply e2etest/pkg/frr/container/container.go:94]: The external containers that act as bgp peers are not related to how we configure FRR as a speaker (as a different BGP mode). Using views in the external containers would add up complexity because we'll need to find different ways to ping/wget/... since the routes won't be installed in the container.

File: e2etest/bgp_test.go (hunk 6 above)
Comment (oribon) [reply e2etest/bgp_test.go:138]: Could you explain why it's confusing? The containersConf describes which containers we want the tests to setup and with which configuration, for them the NeighborConfig is reflected in the bgp conf file (they need to know who are their neighbors), and the RouterConfig is how they should configure/present themselves (reflected in the daemons conf file).

File: e2etest/pkg/frr/container/container.go (hunk 10 above)
Comment (msherif1234) [reply e2etest/pkg/frr/container/container.go:94]: well I was thinking we can use views in speaker configs and we can have two frr external containers ?

**Comment by msherif1234** [comment]
> > How did u test multiple MyASN w/oi using view feature in #1015 ?
> 
> It runs on the native bgp stack - it shows that it works with multiple MyASNs hence we'd want to have these tests pass when FRR with views land. The external containers here don't need to use views (they're using a single ASN).

correct u don't need to use view in external frr routers

File: e2etest/bgp_test.go (hunk 6 above)
Comment (msherif1234) [reply e2etest/bgp_test.go:138]: never mind its clear now Thanks

File: e2etest/pkg/frr/config/config.go (hunk 8 above)
Comment (msherif1234) [reply e2etest/pkg/frr/config/config.go:21]: it seems as of 4.7 the default that network has to exists, I think this need to be debugged further IMHO might be interaction between bgp and rip

File: e2etest/bgp_test.go (hunk 5 above)
Comment (msherif1234) [reply e2etest/bgp_test.go:85]: ok that is clear now thanks

File: e2etest/pkg/frr/config/config.go (hunk 8 above)
Comment (msherif1234) [inline e2etest/pkg/frr/config/config.go:21]: can u add the following tp prepare for Ipv6
```
no bgp default ipv4-unicast
```
I don't know why use next-hop-self it was in the original template and I replaced it with just active

Comment (fedepaol) [reply e2etest/pkg/frr/config/config.go:21]: I think the two things are not related.
`no network import-check` is about being able to advertise networks without having them in the rib.

`no bgp ebgp-requires-policy` is about being able to accept routes when no filters are in place (so it's definetely something needed to add only to the external peer)

File: e2etest/bgp_test.go [hunk 11]
```diff
@@ -21,72 +21,107 @@ package e2e
...
-			framework.ExpectNoError(err)
-			err = frr.StartContainer(frrContainer, frrTestConfigDir)
-			framework.ExpectNoError(err)
+		frrContainers = make([]*frrcontainer.Frr, 0)
+		g := new(errgroup.Group)
+		for _, conf := range containersConf {
+			conf := conf
+			g.Go(func() error {
+				c, err := frrcontainer.New(conf.name, conf.nc, conf.rc, containersNetwork, hostContainerIPv4, hostContainerIPv6)
+				if c != nil {
```
[the remark this answers is not in the document]
Comment (oribon) [reply e2etest/bgp_test.go]: Because Start also calls updateVolumePermissions we could have a situation where a container is created (!=nil) but an error is returned. In that case we should add the container to the slice so it would be stopped.

File: e2etest/bgp_test.go [hunk 12]
```diff
@@ -21,72 +21,107 @@ package e2e
...
-		if skipDockerCmd {
-			if len(frrTestConfigDir) == 0 {
-				framework.Fail("Missing FRR config directory.")
-			}
-		} else {
-			frrTestConfigDir, err = ioutil.TempDir("", "frr-conf")
-			framework.ExpectNoError(err)
-			err = frr.StartContainer(frrContainer, frrTestConfigDir)
-			framework.ExpectNoError(err)
+		frrContainers = make([]*frrcontainer.Frr, 0)
```
[the remark this answers is not in the document]
Comment (oribon) [reply e2etest/bgp_test.go]: done

File: e2etest/pkg/frr/container/container.go [hunk 13]
```diff
@@ -19,20 +20,78 @@ const (
 	frrConfigDir = "config/frr"
 	// FRR routing image.
 	frrImage = "frrouting/frr:v7.5.1"
+	// Host network name.
+	hostNetwork = "host"
 	// FRR container mount destination path.
 	frrMountPath = "/etc/frr"
 )
 
+type Frr struct {
```
[the remark this answers is not in the document]
Comment (oribon) [reply e2etest/pkg/frr/container/container.go]: done

File: e2etest/bgp_test.go [hunk 14]
```diff
@@ -106,21 +150,18 @@ var _ = ginkgo.Describe("BGP", func() {
...
 			framework.ExpectNoError(err)
 
-			pairExternalFRRWithNodes(cs)
+			for _, c := range frrContainers {
+				pairExternalFRRWithNodes(cs, c)
+			}
 
 			allNodes, err = cs.CoreV1().Nodes().List(context.Background(), metav1.ListOptions{})
 			framework.ExpectNoError(err)
+
```
[the remark this answers is not in the document]
Comment (oribon) [reply e2etest/bgp_test.go]: done

File: e2etest/bgp_test.go [hunk 15]
```diff
@@ -21,72 +21,107 @@ package e2e
...
 
 	ginkgo.AfterEach(func() {
 		// Clean previous configuration.
 		err := updateConfigMap(cs, configFile{})
 		framework.ExpectNoError(err)
 
-		if !skipDockerCmd {
-			err = frr.StopContainer(frrContainer, frrTestConfigDir)
-			framework.ExpectNoError(err)
+		g := new(errgroup.Group)
```
[the remark this answers is not in the document]
Comment (oribon) [reply e2etest/bgp_test.go]: done

File: e2etest/pkg/frr/container/container.go (hunk 2 above)
Comment (oribon) [reply e2etest/pkg/frr/container/container.go]: done

File: e2etest/pkg/frr/config/config.go (hunk 8 above)
Comment (oribon) [reply e2etest/pkg/frr/config/config.go:21]: I think it'd be better to keep this PR focused on the current tests, but thanks for thinking ahead :smiley:

**Comment by fedepaol** [comment]
CI is expected to fail, merging this
